include/micm/profiler/instrumentation.hpp

```diff
@@ -0,0 +1,225 @@
+      json << "\"pid\":\"" << id_<< "\",";
+      json << "\"tid\":\"" << result.threadID << "\",";
+      json << "\"ts\":" << result.start.count();
+      json << "}";
+
+      std::lock_guard lock(mutex_);
+      if (current_session_)
+      {
+        output_stream_ << json.str();
+        output_stream_.flush();
+      }
+    }
+
+    static Instrumentor& Get()
+    {
+      static Instrumentor instance;
+      return instance;
+    }
+
+  private:
+    Instrumentor()
+      : current_session_(nullptr)
+    {
+    }
+
+    ~Instrumentor()
+    {
+      EndSession();
+    }
+
+    void WriteHeader()
+    {
+      output_stream_ << "{\"otherData\": {},\"traceEvents\":[{}";
+      output_stream_.flush();
+    }
+
+    void WriteFooter()
+    {
+      output_stream_ << "]}";
+      output_stream_.flush();
+    }
+
+    void InternalEndSession()
+    {
+      if (current_session_)
+      {
+        WriteFooter();
+        output_stream_.close();
+        delete current_session_;
+        current_session_ = nullptr;
+      }
+    }
+
+  private:
+    std::mutex mutex_;
+    InstrumentationSession* current_session_;
+    std::ofstream output_stream_;
+    int id_;
+  };
+
+  class InstrumentationTimer
+  {
+  public:
+    InstrumentationTimer(const char* name)
+      : name_(name), stopped_(false)
+    {
+      start_time_point_ = std::chrono::steady_clock::now();
+    }
+
+    ~InstrumentationTimer()
+    {
+      if (!stopped_)
+        Stop();
+    }
+
+    void Stop()
+    {
+      auto end_time_point = std::chrono::steady_clock::now();
+      auto start = FloatingPointMicroseconds{ start_time_point_.time_since_epoch() };
+      auto elapsed_time = std::chrono::time_point_cast<std::chrono::microseconds>(end_time_point).time_since_epoch() 
+                        - std::chrono::time_point_cast<std::chrono::microseconds>(start_time_point_).time_since_epoch();
+
+      Instrumentor::Get().WriteProfile({name_, start, elapsed_time, std::this_thread::get_id()});
+
+      stopped_ = true;
+    }
+
+  private:
+    const char* name_;
+    std::chrono::time_point<std::chrono::steady_clock> start_time_point_;
+    bool stopped_;
+  };
+
+  namespace InstrumentorUtils
+  {
+    template <size_t N>
+    struct ChangeResult
+    {
+      char Data[N];
+    };
+
+    template <size_t N, size_t K>
+    constexpr auto CleanupOutputString(const char(&expr)[N], const char(&remove)[K])
+    {
+      ChangeResult<N> result = {};
+
+      size_t srcIndex = 0;
+      size_t dstIndex = 0;
+      while (srcIndex < N)
+      {
+        size_t matchIndex = 0;
+        while (matchIndex < K - 1 && srcIndex + matchIndex < N - 1 && expr[srcIndex + matchIndex] == remove[matchIndex])
+          matchIndex++;
+        if (matchIndex == K - 1)
+          srcIndex += matchIndex;
+        result.Data[dstIndex++] = expr[srcIndex] == '"' ? '\'' : expr[srcIndex];
+        srcIndex++;
+      }
+
+      return result;
+    }
+  }
+
+}  // namespace micm
+
+
+#define MICM_PROFILE 1
```

Comment:
```suggestion
```

in `src/CMakeLists.txt`, would you please add something like

```
if (MICM_ENABLE_PROFILE)
    target_compile_definitions(micm PUBLIC MICM_PROFILE)
endif()
```

then the above line wouldn't be needed

Reply:
Good suggestion, thanks!